zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -272,180 +275,149 @@ impl DbFormatChange {
-                {
-                    // Breaking from this loop and dropping the sapling_tree channel
-                    // will cause the sapling compare and delete tasks to finish.
-                    if should_cancel_flag.load(atomic::Ordering::Relaxed) {
-                        break;
-                    }
-
-                    if let Err(error) = sapling_tree_tx.send((height, tree)) {
-                        warn!(?error, "unexpected send error")
-                    }
-                }
-            });
-
-            // Create an unbounded channel for duplicate sapling note commitment tree heights
-            let (unique_sapling_tree_height_tx, unique_sapling_tree_height_rx) = mpsc::channel();
-
-            // Set up task for reading sapling note commitment trees
-            let sapling_compare_task = std::thread::spawn(move || {
-                while let Ok((height, tree)) = sapling_tree_rx.recv() {
-                    let tree = Some(tree);
-                    if prev_sapling_tree != tree {
-                        if let Err(error) = unique_sapling_tree_height_tx.send(height) {
-                            warn!(?error, "unexpected send error")
-                        }
-                        prev_sapling_tree = tree;
-                    }
+            // Prune duplicate Sapling note commitment trees.
+            let mut last_tree = db.sapling_tree_by_height(&Height(0)).expect(
+                "The Sapling note commitment tree for the genesis block should be in the database.",
+            );
+            let mut last_height = Height(1);
+            // Run through all the trees in the finalized chain.
+            for (height, tree) in db.sapling_tree_by_height_range(Height(1)..=initial_tip_height) {
+                // Return early if there is a cancel signal.
+                if !matches!(cancel_receiver.try_recv(), Err(mpsc::TryRecvError::Empty)) {
+                    return;
                 }
-            });
-
-            // Set up task for deleting sapling note commitment trees
-            let db = upgrade_db.clone();
-            let sapling_delete_task = std::thread::spawn(move || {
-                let mut delete_from = sapling_height.next();
-                while let Ok(delete_to) = unique_sapling_tree_height_rx.recv() {
-                    let num_entries = delete_to - delete_from;
-                    if num_entries > 0 {
+                // We delete duplicate trees in batches. A batch is a (possibly empty) series of
+                // identical trees excluding the first tree. We exclude the first tree so that we
+                // don't delete it. We get a batch if we encounter a tree that differs from the
+                // previous one, or if we're at the end of the chain. We get an empty batch if there
+                // are two consecutive differing trees.
+                if last_tree != tree || height == initial_tip_height {
```

Comment:
I think the delete range needs to be different for these two different cases:
1. if the trees aren't equal, we want to keep the tree at `height`
2. if the trees are equal but we've reached the initial tip height, we want to delete the tree at `height`

The code below implements the first case.

Reply:
When I run using all 3 PRs on an old state that was synced to 1719530, I get a panic at the initial tip height:
```
2023-08-25T03:06:36.284392Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate sapling trees after running de-duplicate tree upgrade height=Height(1719530) prev_height=Hei
ght(1719517) tree_root=Root("d6efc94648b492e0a8ee310f4032aa618b565cc124e7cc858137f2f36fdb4e70")
The application panicked (crashed).
Message:  found duplicate sapling or orchard trees after running de-duplicate tree upgrade
Location: zebra-state/src/service/finalized_state/disk_format/upgrade.rs:486

Metadata:
version: 1.1.0+62.g961d45c
Zcash network: Mainnet
running state version: 25.1.1
initial disk state version: 25.0.0
features: default,getblocktemplate_rpcs,metrics_exporter_prometheus,prometheus,release_max_level_info 
branch: state-dedup-fixes
git commit: 961d45c
```

Reply:
This edge case is addressed in [`4b579a6` (#7379)](https://github.com/ZcashFoundation/zebra/pull/7379/commits/4b579a60786751719110aee6fcfdde9262b02aef).